### PR TITLE
docs: add replacement for 'ENCODER_BY_TYPE' to migration guide

### LINKED ISSUE #8426: Cannot import ENCODERS_BY_TYPE
### Initial Checks

- [X] I confirm that I'm using Pydantic V2

### Description

I am migrating my code from V1 to V2. In my V1 code, I modified ENCODERS_BY_TYPE to turn bson ObjectIds. I can no longer import ENCODERS_BY_TYPE, how should I fix this issue?

### Example Code

```Python
from pydantic.json import ENCODERS_BY_TYPE
from bson import ObjectId
ENCODERS_BY_TYPE[ObjectId] = lambda v: str(v)
```

### Python, Pydantic & OS Version

```Text
"
             pydantic version: 2.5.2
        pydantic-core version: 2.14.5
          pydantic-core build: profile=release pgo=true
                 install path: /home/tom/miniconda3/envs/routingdeploy2/lib/python3.10/site-packages/pydantic
               python version: 3.10.0 (default, Mar  3 2022, 09:58:08) [GCC 7.5.0]
                     platform: Linux-6.1.11-76060111-generic-x86_64-with-glibc2.35
             related packages: email-validator-2.0.0.post2 typing_extensions-4.8.0 pydantic-settings-2.1.0 fastapi-0.105.0
```

### PR DESCRIPTION
**Pull request by bL34cHig0**
## Change Summary
I appended a work around code for replacing `ENCODERS_BY_TYPE` in V2 to the docs **Migration Guide**

## Related issue number
fix #8426

## Checklist

* [x] The pull request title is a good summary of the changes - it will be used in the changelog
* [ ] Unit tests for the changes exist
* [ ] Tests pass on CI
* [x] Documentation reflects the changes where applicable
* [x] My PR is ready to review, **please add a comment including the phrase "please review" to assign reviewers**

Selected Reviewer: @davidhewitt

### CODE REVIEW

**Comment by bL34cHig0** [comment]
@sydney-runkle 

please review my pull request and I look forward to your feedback. Thanks :)

**Review by sydney-runkle — CHANGES_REQUESTED** [review]
Alright, looking at this more closely, I think there are a few things we should change.

First, let's add `pydantic.json.ENCODERS_BY_TYPE` to the list of deprecated things in V2.

[This section](https://docs.pydantic.dev/latest/migration/#changes-to-pydanticbasemodel) of the existing migration guide is relevant:

* We have significantly expanded Pydantic's capabilities related to customizing serialization. In particular, we have added the [@field_serializer](https://docs.pydantic.dev/latest/api/functional_serializers/#pydantic.functional_serializers.field_serializer), [@model_serializer](https://docs.pydantic.dev/latest/api/functional_serializers/#pydantic.functional_serializers.model_serializer), and [@computed_field](https://docs.pydantic.dev/latest/api/fields/#pydantic.fields.computed_field) decorators, which each address various shortcomings from Pydantic V1.
  * See [Custom serializers](https://docs.pydantic.dev/latest/concepts/serialization/#custom-serializers) for the usage docs of these new decorators.
* Due to performance overhead and implementation complexity, we have now deprecated support for specifying json_encoders in the model config. This functionality was originally added for the purpose of achieving custom serialization logic, and we think the new serialization decorators are a better choice in most common scenarios.

Thus, I think instead of the current alternative we have listed, we can probably just make a note about how you can use a custom field serializer for the given type instead.

Let me know if this makes sense and if you have any follow up questions :).

**Comment by sydney-runkle** [comment]
Also see https://docs.pydantic.dev/latest/api/config/#pydantic.config.ConfigDict.json_encoders

**Comment by bL34cHig0** [comment]
> @bL34cHig0,
> 
> Thanks a ton for your contribution. We're looking forward to this docs update.
> 
> Alright, looking at this more closely, I think there are a few things we should change.
> 
> First, let's add `pydantic.json.ENCODERS_BY_TYPE` to the list of deprecated things in V2.
> 
> [This section](https://docs.pydantic.dev/latest/migration/#changes-to-pydanticbasemodel) of the existing migration guide is relevant:
> 
>     * We have significantly expanded Pydantic's capabilities related to customizing serialization. In particular, we have added the [@field_serializer](https://docs.pydantic.dev/latest/api/functional_serializers/#pydantic.functional_serializers.field_serializer), [@model_serializer](https://docs.pydantic.dev/latest/api/functional_serializers/#pydantic.functional_serializers.model_serializer), and [@computed_field](https://docs.pydantic.dev/latest/api/fields/#pydantic.fields.computed_field) decorators, which each address various shortcomings from Pydantic V1.
>       
>       * See [Custom serializers](https://docs.pydantic.dev/latest/concepts/serialization/#custom-serializers) for the usage docs of these new decorators.
> 
>     * Due to performance overhead and implementation complexity, we have now deprecated support for specifying json_encoders in the model config. This functionality was originally added for the purpose of achieving custom serialization logic, and we think the new serialization decorators are a better choice in most common scenarios.
> 
> 
> Thus, I think instead of the current alternative we have listed, we can probably just make a note about how you can use a custom field serializer for the given type instead.
> 
> Let me know if this makes sense and if you have any follow up questions :).

 @sydney-runkle

Thanks for the feedback and I apologize for the late response.

However, I have a couple of follow up questions I'd like you to address:

- You mentioned we should make a note about how one can use a custom field serializer for the given type instead. Are you referring to the [field_serializer](https://docs.pydantic.dev/latest/api/functional_serializers/#pydantic.functional_serializers.field_serializer)?
- If so, do I replace the work around code with an example of how to use a custom field serializer for the given type or do I add it instead?

Thanks again :)

**Comment by sydney-runkle** [comment]
@bL34cHig0,

Thanks for your work, no need to apologize!

> You mentioned we should make a note about how one can use a custom field serializer for the given type instead. Are you referring to the [field_serializer](https://docs.pydantic.dev/latest/api/functional_serializers/#pydantic.functional_serializers.field_serializer)?

A field serializer would work here, but I think an annotated type with a plain serializer might work even better. Here's an [example](https://docs.pydantic.dev/latest/concepts/serialization/#custom-serializers) of something like that.

> If so, do I replace the work around code with an example of how to use a custom field serializer for the given type or do I add it instead?

Right, I'd replace the workaround code. We don't want to suggest using `json_encoders`, as that'll be deprecated in the future. Thus, there's no perfect equivalent in V2, but we should at least mention that `ENCODER_BY_TYPE` is no longer available.

**Comment by bL34cHig0** [comment]
@sydney-runkle 

Thanks for clearing things up. I'll work on it.

File: docs/migration.md [hunk 1]
```diff
@@ -779,29 +779,30 @@ For `ConstrainedStr` you can use [`StringConstraints`][pydantic.types.StringCons
...
+| Pydantic V1                                | Pydantic V2                                           |
+|--------------------------------------------|-------------------------------------------------------|
+| `pydantic.tools.schema_of`                 | `pydantic.deprecated.tools.schema_of`                 |
+| `pydantic.tools.parse_obj_as`              | `pydantic.deprecated.tools.parse_obj_as`              |
+| `pydantic.tools.schema_json_of`            | `pydantic.deprecated.tools.schema_json_of`            |
+| `pydantic.json.pydantic_encoder`           | `pydantic.deprecated.json.pydantic_encoder`           |
+| `pydantic.validate_arguments`              | `pydantic.deprecated.decorator.validate_arguments`    |
+| `pydantic.json.custom_pydantic_encoder`    | `pydantic.deprecated.json.custom_pydantic_encoder`    |
+| `pydantic.json.timedelta_isoformat`        | `pydantic.deprecated.json.timedelta_isoformat`        |
+| `pydantic.json.ENCODERS_BY_TYPE`           | `pydantic.deprecated.json.ENCODERS_BY_TYPE`           |
```
Comment (sydney-runkle) [inline docs/migration.md:791]: Looks good overall, but no need to change the formatting. Should be a 1 line diff. Happy to accept once you change things back 👍

**Review by sydney-runkle — CHANGES_REQUESTED** [review]
One quick change request, thanks!

File: docs/migration.md (hunk 1 above)
Comment (bL34cHig0) [reply docs/migration.md:791]: Oh my bad, it was inadvertent. Will correct it now.

**Review by sydney-runkle — APPROVED** [review]
Looks great, thanks for adding this :)

**Comment by bL34cHig0** [comment]
> Looks great, thanks for adding this :)

Don't mention it :) 

I'll add a note about how one can use a custom field serializer for the given type soon. Thanks 👍

**Comment by mcleantom** [comment]
I feel like the documentation title and example shouldnt just say it is for "ObjectId" fields, as it can be for any field that doesnt have a default json encoding?

**Pull request merged**